crates/wasmparser/src/validator/types.rs

```diff
@@ -392,6 +404,16 @@ impl ModuleType {
         self.imports.get(&(module, name) as &dyn ModuleImportKey)
     }
 
+    pub(crate) fn to_instance_type(&self) -> ModuleInstanceType {
+        ModuleInstanceType {
+            type_size: self
+                .exports
+                .iter()
+                .fold(1, |acc, (_, ty)| acc + ty.type_size()),
+            exports: self.exports.clone(),
+        }
```

Comment:
Instead of cloning things around here, I think something that the old module linking implementation did was to basically say "this instance is an instantaition of this module" which means we'd chase a pointer in a sense but I think would avoid having to clone anything here related to the type information.

Reply:
I've removed the clone calls in favor of storing a reference (via a `TypeId`) to the underlying module or component type for the instance.